### ISSUE TITLE
MongoDB Unfrezable type Error

### ISSUE THREAD

**Issue by kimus** [metabase/metabase#11121]
**Describe the bug**
When creating the 'Question' everything runs has expected. But
throws an error after saving the 'Question': `Unfrezable type: class.bson.types.ObjectId`

I Can, again, edit the 'Question' change something and filter, again with success.
But after saving, and refresh the 'Question' the same error is displayed.

<img width="643" alt="image" src="https://user-images.githubusercontent.com/188404/66643966-c2e98e80-ec17-11e9-93d6-193c414bb9f4.png">

Using metabase 0.33.4 but same problem have occurred in 0.32.x

**Comment by kimus** [comment]
Here it's the error that popups in the logs:
```
[35cfd9a6-e3ff-4a09-bf8e-b69b55d1033f] 2019-10-14T16:55:20+01:00 WARN metabase.query-processor.middleware.process-userland-query Query failure {:status :failed,
 :class clojure.lang.ExceptionInfo,
 :error "Unfreezable type: class org.bson.types.ObjectId",
 :stacktrace
 ("taoensso.nippy$throw_unfreezable.invokeStatic(nippy.clj:720)"
  "taoensso.nippy$throw_unfreezable.invoke(nippy.clj:718)"
  "taoensso.nippy$fn__27535.invokeStatic(nippy.clj:924)"
  "taoensso.nippy$fn__27535.invoke(nippy.clj:905)"
  "taoensso.nippy$fn__27338$G__27333__27345.invoke(nippy.clj:314)"
  "taoensso.nippy$fn__27373.invokeStatic(nippy.clj:331)"
  "taoensso.nippy$fn__27373.invoke(nippy.clj:316)"
  ... removed ...
  "--> query_processor.middleware.cache_backend.db$compress_until_max$fn__68723.invoke(db.clj:70)"
  "query_processor.middleware.cache_backend.db$compress_until_max.invokeStatic(db.clj:67)"
  "query_processor.middleware.cache_backend.db$compress_until_max.invoke(db.clj:60)"
  "query_processor.middleware.cache_backend.db$save_results_BANG_.invokeStatic(db.clj:84)"
  "query_processor.middleware.cache_backend.db$save_results_BANG_.invoke(db.clj:77)"
  "query_processor.middleware.cache_backend.db$reify__68729.save_results_BANG_(db.clj:101)"
  "query_processor.middleware.cache$save_results_BANG_.invokeStatic(cache.clj:88)"
  "query_processor.middleware.cache$save_results_BANG_.invoke(cache.clj:86)"
  "query_processor.middleware.cache$save_results_if_successful_BANG_.invokeStatic(cache.clj:103)"
  "query_processor.middleware.cache$save_results_if_successful_BANG_.invoke(cache.clj:97)"
  "query_processor.middleware.cache$run_query_with_cache$respond__37269.invoke(cache.clj:114)"
  "query_processor.middleware.async$async__GT_sync$fn__20798.invoke(async.clj:23)"
  "query_processor.middleware.async_wait$runnable$fn__36561.invoke(async_wait.clj:89)"),
 :query
 {:constraints {:max-results 10000, :max-results-bare-rows 2000},
  :type :query,
  :middleware {:userland-query? true},
  :info
  {:executed-by 1,
   :context :question,
   :card-id 575,
   :dashboard-id nil,
   :query-hash [48, 43, -3, -83, 61, 72, 47, 19, 110, -117, 104, -65, 83, 40, 120, -68, -79, -56, 81, 39, 1, -128, -100, -75, -36, -24, 24, 103, 80, -123, -38, -19]},
  :query
  {:source-table 61,
   :filter [:and [:time-interval [:field-id 1860] :current :month] [:time-interval [:field-id 695] :current :month] [:= [:field-id 1781] false]],
   :aggregation [[:sum [:field-id 721]]],
   :breakout [[:field-id 690]],
   :order-by [[:desc [:aggregation 0]]],
   :limit 1},
  :parameters [],
  :async? true,
  :cache-ttl 30},
 :preprocessed
   ... removed ...
 :native
 {:projections [:customerId :sum],
  :query
  [{"$project"
    ... removed ...
  :mbql? true},
 :ex-data {:type org.bson.types.ObjectId, :as-str "#object[org.bson.types.ObjectId 0x44f7236e \"5d9dea17465dcb00dfdd583a\"]"}}
```

**Comment by kimus** [comment]
This is related with cache. With cache disabled the output works has expected

**Comment by flamber** [comment]
@kimus Which database+version are you using for metadata?

**Comment by kimus** [comment]
@flamber the logs are from 0.32 but this also happens in 0.33

**Comment by flamber** [comment]
@kimus I'm asking which database you're using for Metabase's application data (aka metadata).
If you're unsure, then copy/paste the Diagnostic Info from Admin > Troubleshooting.

**Comment by kimus** [comment]
@flamber sorry, I'm using the default docker installation with H2 database

```javascript
{
  "browser-info": {
    "language": "en-US",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:70.0) Gecko/20100101 Firefox/70.0",
    "vendor": ""
  },
  "system-info": {
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.4+11",
    "java.vendor": "AdoptOpenJDK",
    "java.vendor.url": "https://adoptopenjdk.net/",
    "java.version": "11.0.4",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.4+11",
    "os.name": "Linux",
    "os.version": "4.4.0-165-generic",
    "user.language": "en",
    "user.timezone": "GMT"
  },
  "metabase-info": {
    "databases": [
      "mongo"
    ],
    "hosting-env": "unknown",
    "application-database": "h2",
    "run-mode": "prod",
    "version": {
      "date": "2019-10-07",
      "tag": "v0.33.4",
      "branch": "release-0.33.x",
      "hash": "9559406"
    },
    "settings": {
      "report-timezone": "Europe/London"
    }
  }
}
```

**Comment by kimus** [comment]
This issue is greatly impacting our production server because we had to disable metabase cache. We have several heavy queries in our setup and it's hard to control the 'abuse' of dashboard refreshing from the users that was minimized by caching.

Is it possible to know what would be the timeline for resolving this issue? Thanks.

**Comment by flamber** [comment]
@kimus I would highly recommend that you [migrate away from H2](https://www.metabase.com/docs/latest/operations-guide/migrating-from-h2.html). I don't think it will fix your problem, but it will prevent you from ending up with a corrupted H2 at some point.

Seems like the error comes from Nippy, when it tries to serialize unknown types:
https://stackoverflow.com/questions/49668404/metabase-clojure-error-unfreezable-type-class-org-postgresql-jdbc-pgarray

As for a timeline for a fix, I don't know - it's the first time it's been reported. But I can see Simon has assigned himself, so that usually means he's looking into this.

**Comment by kimus** [comment]
@sbelak just remembering that this is very important for me and my team. All dashboards give errors if cache is enabled. This is not a minor query/gui issue. This is impacting all the queries used by the dashboards. And because the cache needs to be disabled, it's impacting the database/server performance.

**Comment by tomahock** [comment]
+1

**Comment by kimus** [comment]
Upgraded Metabase to 0.33.7.3 and still the same errors!!

**Comment by flamber** [comment]
@kimus This will be fixed on the upcoming 0.34.0 - the fix is included in 0.34.0-RC1, you can try it out:
https://github.com/metabase/metabase/releases/tag/0.34.0-rc1

**Comment by vmercierfr** [comment]
We'll have the issue in 0.34.2 (with PostgreSQL connector).

**Comment by flamber** [comment]
@vmercierfr I would recommend that you open a new issue, fill out all the information and then reference this issue too. Since this issue was specific to MongoDB and fixed on 0.34.0
